File: python/import.py
import flare
import json

def importData():
	with open("model.json") as json_file:
		data = json.load(json_file)
		
		for environment in data.get('environments', []):
			importEnvironment(environment)

def importEnvironment(source):
	environment = {} 
	environment['name'] = source.get('name', 'untitled')
	environment['description'] = source.get('description', '')
	environment['geofence'] = source.get('geofence', {})
	environment['perimeter'] = source.get('perimeter', {})
	environment['data'] = source.get('data', {})
	environment['angle'] = source.get('angle', 0)
	
	result = flare.newEnvironment(environment)
	environmentId = result['_id']
	print('Imported environment', result['name'])
	
	for zone in source.get('zones', []):
		importZone(environmentId, zone)

	for device in source.get('devices', []):
		importDevice(environmentId, device)

def importZone(environmentId, source):
	zone = {}
	zone['name'] = source.get('name', 'untitled')
	zone['description'] = source.get('description', '')
	zone['perimeter'] = source.get('perimeter', {})
	zone['data'] = source.get('data', {})
	
	result = flare.newZone(environmentId, zone)
	zoneId = result['_id']
	print('Imported zone', result['name'])

	for thing in source.get('things', []):
		importThing(environmentId, zoneId, thing)

def importThing(environmentId, zoneId, source):
	thing = {}
	thing['name'] = source.get('name', 'untitled')
	thing['description'] = source.get('description', '')
	thing['position'] = source.get('position', {'x': 0, 'y': 0})
	thing['data'] = source.get('data', {})
	
	result = flare.newThing(environmentId, zoneId, thing)
	print('Imported thing', result['name'])

def importDevice(environmentId, source):
	device = {}
	device['name'] = source.get('name', 'untitled')
	device['description'] = source.get('description', '')
	device['position'] = source.get('position', {'x': 0, 'y': 0})
	device['data'] = source.get('data', {})
	
	result = flare.newDevice(environmentId, device)
	print('Imported device', result['name'])

importData()

File: server/import.js
// Flare Importer
// Andrew Zamler-Carhart

var flare = this;
var flaredb = require("./flaredb");
var modelPath = process.argv[2] || 'public/home/Home.json';
var model = require('./' + modelPath);
var tasks = 0;

/*
flaredb.Environment.remove({});
flaredb.Zone.remove({});
flaredb.Thing.remove({});
flaredb.Device.remove({});
*/

console.log('Importing from ' + modelPath);

for (var i = 0; i < model.environments.length; i++) {
	createEnvironment(model.environments[i]);
};

function createEnvironment(environment) {
	environment.created = new Date();
	environment.modifed = new Date();

	start();
	flaredb.Environment.create(environment, function (err, newEnvironment) {
		if (err) console.log(err);
		console.log('Environment ' + newEnvironment._id + ': ' + newEnvironment.name);

		for (var j = 0; j < environment.zones.length; j++) {
			createZone(newEnvironment._id, environment.zones[j]);
		}

		if (environment.devices != undefined) {
			for (var k = 0; k < environment.devices.length; k++) {
				createDevice(newEnvironment._id, environment.devices[k]);
			}
		}

		finish();
	});
}

function createZone(environment_id, zone) {
	zone.environment = environment_id;
	zone.created = new Date();
	zone.modifed = new Date();

	start();
	flaredb.Zone.create(zone, function (err, newZone) {
		if (err) console.log(err);
		console.log('Zone ' + newZone._id + ': ' + newZone.name);

		for (var l = 0; l < zone.things.length; l++) {
			createThing(environment_id, newZone._id, zone.things[l]);
		}
		
		finish();
	});
}

function createThing(environment_id, zone_id, thing) {
	thing.environment = environment_id;
	thing.zone = zone_id;
	thing.created = new Date();
	thing.modifed = new Date();
	
	start();
	flaredb.Thing.create(thing, function (err, newThing) {
		if (err) console.log(err);
		console.log('Thing ' + newThing._id + ': ' + newThing.name);

		finish();
	});
}

function createDevice(environment_id, device) {
	device.environment = environment_id;
	device.created = new Date();
	device.modifed = new Date();
	
	start();
	flaredb.Device.create(device, function (err, newDevice) {
		if (err) console.log(err);
		console.log('Device ' + newDevice._id + ': ' + newDevice.name);
		
		finish();
	});
}

function start() {
	tasks++;
}

function finish() {
	tasks--;
	if (tasks == 0) {
		console.log('Done.');
		process.exit(0);
	}
}
